src/ros_filter.cpp

```diff
@@ -1975,27 +1997,39 @@ namespace RobotLocalization
       for (size_t j = 0; j < dimension; j++)
       {
         covariance(i, j) = arr[dimension * i + j];
+      }
+    }
 
+    for (size_t i = 0; i < dimension; i++)
+    {
+      for (size_t j = 0; j < dimension; j++)
+      {
         if (printDiagnostics_)
         {
-          std::string iVar = stateVariableNames_[offset + i];
-
-          if (covariance(i, j) > 1e3 && (updateVector[offset  + i] || updateVector[offset  + j]))
+          if (previousCovariances_.count(topicName) > 0)
           {
-            std::string jVar = stateVariableNames_[offset + j];
+            const Eigen::MatrixXd& previousCovariance = previousCovariances_[topicName];
 
-            std::stringstream stream;
-            stream << "The covariance at position (" << dimension * i + j << "), which corresponds to " <<
-                (i == j ? iVar + " variance" : iVar + " and " + jVar + " covariance") <<
-                ", the value is extremely large (" << covariance(i, j) << "), but the update vector for " <<
-                (i == j ? iVar : iVar + " and/or " + jVar) << " is set to true. This may produce undesirable results.";
+            const double previousVolume = computeCovarianceVolume(previousCovariance);
+            const double volume = computeCovarianceVolume(covariance);
 
-            addDiagnostic(diagnostic_msgs::DiagnosticStatus::WARN,
-                          topicName + "_covariance",
-                          stream.str(),
-                          false);
+            if (volume - previousVolume > 1e2)
+            {
+              std::stringstream stream;
+              stream << "The covariance volume (" << volume <<
+                  ") has grown too much wrt the previous volume (" << previousVolume <<
+                  "). This may produce undesirable results.";
+
+              addDiagnostic(diagnostic_msgs::DiagnosticStatus::WARN,
+                            topicName + "_covariance",
+                            stream.str(),
+                            false);
+            }
```

Comment:
Knowing EKFs as we do, this message makes sense. For your every day user, though, we might simply want to state that the covariance experienced massive growth between time steps.